### PR TITLE
Add documentation for disk based eviction

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Documents new features in eviction.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1196)

### CODE REVIEW

**Comment by devin-donnelly** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-:-KRFLyXf5DZ-bZppc7kU:bnfp4nl)*

File: docs/admin/out-of-resource.md [hunk 1]
```diff
@@ -31,10 +31,25 @@ summary API.
...
 |------------------|---------------------------------------------------------------------------------|
 | `memory.available` | `memory.available` := `node.status.capacity[memory]` - `node.stats.memory.workingSet` |
+| `nodefs.available` | `nodefs.available` := `node.stats.fs.available` |
+| `nodefs.inodesFree` | `nodefs.inodesFree` := `node.stats.fs.inodesFree` |
+| `imagefs.available` | `imagefs.available` := `node.stats.runtime.imagefs.available` |
+| `imagefs.inodesFree` | `imagefs.inodesFree` := `node.stats.runtime.imagefs.inodesFree` |
 
-In future releases, the `kubelet` will support the ability to trigger eviction decisions based on disk pressure.
+Each of the above signals support either a literal or percentage based value.  The percentage based value
+is calculated relative to the total capacity associated with each signal.
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: Does this hold for `memory.available` too?

File: docs/admin/out-of-resource.md [hunk 2]
```diff
@@ -31,10 +31,25 @@ summary API.
...
+
+1. The `nodefs` filesystem that kubelet uses for volumes, daemon logs, etc.
+1. The `imagefs` filesystem that container runtimes uses for storing images and container writable layers.
+
+`imagefs` is optional. `kubelet` auto-discovers these filesystems using cAdvisor.  `kubelet` does not care about any 
+other filesystems. Any other types of configurations are not currently supported by the kubelet. For example, it is
+*not OK* to store volumes and logs in a dedicated `imagefs`.
+
+In future releases, the `kubelet` will deprecate the existing [garbage collection](/docs/admin/garbage-collection/)
+support in favor of eviction in response to disk pressure.
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: We'll still do some garbage collection though, right? Saying we're going to deprecate it is confusing.

File: docs/admin/out-of-resource.md [hunk 3]
```diff
@@ -47,6 +62,14 @@ Each threshold is of the following form:
...
 * valid `operator` tokens are `<`
 * valid `quantity` tokens must match the quantity representation used by Kubernetes
+* an eviction threshold can be expressed as a percentage if ends with `%` token.
+
+For example, if a node has `10Gi` of memory, and the desire is to induce eviction
+if available memory falls below `1Gi`, an eviction threshold can be specified as either
+of the following (but not both).
+
+* `memory.available<10%`
+* `memory.available<1Gi`
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: Where are these specified? I assume there's a flag / config for it somewhere? What are the defaults?

File: docs/admin/out-of-resource.md [hunk 4]
```diff
@@ -151,6 +175,76 @@ and the node only has `Guaranteed` pod(s) remaining, then the node must choose t
...
+
+#### Ranking pods for eviction
+
+Local disk is a `BestEffort` resource.  If necessary, `kubelet` will evict pods one at a time to reclaim
+disk when `DiskPressure` is encountered.  The `kubelet` will rank pods by quality of service.  If the `kubelet`
+is responding to `inode` starvation, it will reclaim `inodes` by evicting pods with the lowest quality of service
+first.  If the `kubelet` is responding to lack of available disk, it will rank pods within a quality of service
+that consumes the largest amount of disk and kill those first.
+
+##### Without Imagefs
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: (nit) use the same with/without ordering as above

File: docs/admin/out-of-resource.md [hunk 5]
```diff
@@ -151,6 +175,76 @@ and the node only has `Guaranteed` pod(s) remaining, then the node must choose t
...
+- local volumes + logs & writable layer of all its containers.
+
+##### With Imagefs
+
+If `nodefs` is triggering evictions, `kubelet` will sort pods based on the usage on `nodefs`
+- local volumes + logs of all its containers.
+
+If `imagefs` is triggering evictions, `kubelet` will sort pods based on the writable layer usage of all its containers.
+
+##### Minimum eviction reclaim
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: Is this specific to disk-based evictions?

File: docs/admin/out-of-resource.md [hunk 6]
```diff
@@ -223,3 +318,42 @@ candidate set of pods provided to the eviction strategy.
...
+in favor of the simpler configuation supported around eviction.
+
+| Existing Flag | New Flag | Rationale |
+| ------------- | -------- | --------- |
+| `--image-gc-high-threshold` | `--eviction-hard` or `eviction-soft` | existing eviction signals can capture image garbage collection |
+| `--image-gc-low-threshold` | `--eviction-minimum-reclaim` | eviction reclaims achieve the same behavior |
+| `--maximum-dead-containers` | | deprecated once old logs are stored outside of container's context |
+| `--maximum-dead-containers-per-container` | | deprecated once old logs are stored outside of container's context |
+| `--minimum-container-ttl-duration` | | deprecated once old logs are stored outside of container's context |
+| `--low-diskspace-threshold-mb` | `--eviction-hard` or `eviction-soft` | this use case is better handled by this proposal |
```
Comment (timstclair) [inline docs/admin/out-of-resource.md]: (nit) I don't think this is a proposal anymore?

**Comment by timstclair** [comment]
Review pass complete.

File: docs/admin/out-of-resource.md (hunk 2 above)
Comment (yujuhong) [inline docs/admin/out-of-resource.md]: Garbage collection (for both image and containers) is more aggressive than the disk eviction AFAIK. For example, it deletes containers associated with deleted pods, and it also ensure we kept no more than N containers per (pod, container) tuple. Is the plan to fold garbage collection into eviction?

File: docs/admin/out-of-resource.md [hunk 7]
```diff
@@ -223,3 +318,42 @@ candidate set of pods provided to the eviction strategy.
...
+
+### kubelet may evict more pods than needed
+
+The pod eviction may evict more pods than needed due to stats collection timing gap. This can be mitigated by adding
+the ability to get root container stats on an on-demand basis (https://github.com/google/cadvisor/issues/1247) in the future.
+
+### How kubelet ranks pods for eviction in response to inode exhaustion
+
+At this time, it is not possible to know how many inodes were consumed by a particular container.  If the `kubelet` observes
+inode exhaustion, it will evict pods by ranking them by quality of service.  The following issue has been opened in cadvisor
```
Comment (yujuhong) [inline docs/admin/out-of-resource.md]: How does it rank pods in the same QoS class?

File: docs/admin/out-of-resource.md [hunk 8]
```diff
@@ -223,3 +318,42 @@ candidate set of pods provided to the eviction strategy.
...
 create `BestEffort` pods to avoid being identified as a candidate pod
 for eviction. Instead `DaemonSet` should ideally launch `Guaranteed` pods.
+
+## Deprecation of existing feature flags to reclaim disk
+
+`kubelet` has been freeing up disk space on demand to keep the node stable.
+
+As disk based eviction matures, the following `kubelet` flags will be marked for deprecation
+in favor of the simpler configuation supported around eviction.
+
```
Comment (yujuhong) [inline docs/admin/out-of-resource.md]: I like simplifying the configuration, just one question. Will eviction only happen when the threshold is met? More aggressive, periodic cleanup, seems desirable IMO.

**Comment by vishh** [comment]
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/admin/out-of-resource.md, line 40 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-40:-KRVTWQ2Py-FTknxXNNO:b-ry0ney) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L40)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Does this hold for `memory.available` too?
</details>
My understanding is yes. Relevant code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L183)

---

*[docs/admin/out-of-resource.md, line 49 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVTqy8IRBj_-VNL5d8:-KRVTqy8IRBj_-VNL5d9:bdwql29) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L49)):*
> ```Markdown
> `imagefs` is optional. `kubelet` auto-discovers these filesystems using cAdvisor.  `kubelet` does not care about any 
> other filesystems. Any other types of configurations are not currently supported by the kubelet. For example, it is
> *not OK* to store volumes and logs in a dedicated `imagefs`.
> ```

nit: dedicated `filesystem` instead of `imagefs`.

---

*[docs/admin/out-of-resource.md, line 52 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-52:-KRVTyguvE4Avrn4IyU7:b-f8dxda) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L52)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Garbage collection (for both image and containers) is more aggressive than the disk eviction AFAIK. For example, it deletes containers associated with deleted pods, and it also ensure we kept no more than N containers per (pod, container) tuple. Is the plan to fold garbage collection into eviction?
</details>
Yes. The plan is to only expose evictions as the feature to end users. GC will be an internal implementation detail.

---

*[docs/admin/out-of-resource.md, line 72 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-72:-KRVU5p3EnJftsaF1Il_:b-qwkvtm) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L72)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Where are these specified? I assume there's a flag / config for it somewhere? What are the defaults?
</details>
--eviction-hard is defaulted [here](https://github.com/kubernetes/kubernetes/blob/7523669699be96c49faddc3e7be83d2eae56360e/pkg/apis/componentconfig/v1alpha1/defaults.go#L331).
In v1.5 disk evictions will also become opt-out in upstream.
Meanwhile, k8s clusters on GCE/GKE already have disk evictions as opt-opt. 

---

*[docs/admin/out-of-resource.md, line 181 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVUiTEoxcDNh-dumrv:-KRVUiTEoxcDNh-dumrw:b-lckirq) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L181)):*
> ```Markdown
> 
> The `kubelet` responds differently to disk based eviction thresholds if the machine has a
> dedicatd `imagefs` configured for the container runtime.
> ```

nit: typo - dedicated

---

*[docs/admin/out-of-resource.md, line 223 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-223:-KRVV3Ldh1mciUm_acsm:bglpzzy) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L223)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Is this specific to disk-based evictions?
</details>
Nope. +1 for moving to a separate section

---

*[docs/admin/out-of-resource.md, line 247 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVVtbwHLCC9ASQiPB-:-KRVVtbx-GUYPXnAawIM:b-v1w2sg) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L247)):*
> ```Markdown
> 
> The default `eviction-minimum-reclaim` is `0` for all resources.
> ```

What about documentation on inodes based eviction behavior?

---

*[docs/admin/out-of-resource.md, line 328 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-328:-KRVW1CNgX8lRcceGwxw:b-p4j4at) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L328)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> I like simplifying the configuration, just one question. Will eviction only happen when the threshold is met? More aggressive, periodic cleanup, seems desirable IMO. 
</details>
I assume you are referring to container GC. I wish we had this discussion on a k8s issue given that this feature is already in place. 
Container GC will eventually happen all the time once logs are pulled out of dead containers and kubelet can checkpoint outside of containers. 
If you were referring to image GC, there is really no need to GC images unless there is disk pressure. Image locality is helpful for pod startup latency. 

---

*[docs/admin/out-of-resource.md, line 336 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-336:-KRVWiebRDxCTTwKxdW2:b-xh0r97) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L336)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> (nit) I don't think this is a proposal anymore?
</details>
I don't think the rationale column is necessary here for the documentation. 

---

*[docs/admin/out-of-resource.md, line 346 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVcw2oEkYd7tvEjtpo:-KRVcw2oEkYd7tvEjtpp:b-31kvcy) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L346)):*
> ```Markdown
> increases within that window rapidly, the `kubelet` may not observe `MemoryPressure` fast enough, and the `OOMKiller`
> will still be invoked.  We intend to integrate with the `memcg` notification API in a future release to reduce this
> latency, and instead have the kernel tell us when a threshold has been crossed immmediately.
> ```

It might be worth mentioning a workaround. For example, set eviction thresholds to <= 75% for example to reduce the occurrence of this issue.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196)*

**Comment by devin-donnelly** [comment]
@derekwaynecarr , have you had a chance to address the comments from @vishh and @timstclair yet?

**Comment by derekwaynecarr** [comment]
@devin-donnelly - I am addressing feedback now.

File: docs/admin/out-of-resource.md (hunk 1 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: yes.  it holds for all the things.

File: docs/admin/out-of-resource.md (hunk 2 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: yes

File: docs/admin/out-of-resource.md (hunk 3 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: they are specified as flags on the kubelet that are defined later in this doc.

this document does not define defaults.

defaults right now are specific to a commercial offering imo.

File: docs/admin/out-of-resource.md (hunk 4 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: ack

File: docs/admin/out-of-resource.md (hunk 5 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: no, its supported for any resource under pressure, but it was new to 1.4 so got collapsed into this PR.  the heading depth confuses this, so i updated it, thanks.

File: docs/admin/out-of-resource.md (hunk 6 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: ack, removed.

File: docs/admin/out-of-resource.md (hunk 7 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: this is addressed under "Known issues"  section "How kubelet ranks pods for eviction in response to inode exhaustion", in the future, we want it to rank by consumption when we can know it.

Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: to be clear, it does not differentiate within a class, so its random.

File: docs/admin/out-of-resource.md (hunk 8 above)
Comment (derekwaynecarr) [inline docs/admin/out-of-resource.md]: we should discuss this in a design issue, its tangental to the doc.  what's documented is the current plan of record.

**Comment by derekwaynecarr** [comment]
*[docs/admin/out-of-resource.md, line 72 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-72:-KRoSJ49nYOxYFjFcHe7:b-7cy4gr) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L72)):*
<details><summary><i>Previously, derekwaynecarr (Derek Carr) wrote…</i></summary>
> they are specified as flags on the kubelet that are defined later in this doc.
> 
> 
> 
> this document does not define defaults.
> 
> 
> 
> defaults right now are specific to a commercial offering imo.
</details>
actually, we do default eviction on memory.  let me update this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196)*

**Comment by derekwaynecarr** [comment]
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/admin/out-of-resource.md, line 49 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVTqy8IRBj_-VNL5d8:-KRoS9sfm23TkE44xHeL:b-qssb2m) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L49)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> nit: dedicated `filesystem` instead of `imagefs`.
</details>
fixing

---

*[docs/admin/out-of-resource.md, line 247 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVVtbwHLCC9ASQiPB-:-KRoTCR9f56-Qis8TOwS:bxrwhby) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L247)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> What about documentation on inodes based eviction behavior?
</details>
i give details under known issues.

---

*[docs/admin/out-of-resource.md, line 346 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KRVcw2oEkYd7tvEjtpo:-KRoURj5lk_lfQee4Ny6:bq3jr6k) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L346)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> It might be worth mentioning a workaround. For example, set eviction thresholds to <= 75% for example to reduce the occurrence of this issue.
</details>
added text.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196)*

**Comment by derekwaynecarr** [comment]
@vishh @timstclair -- all comments addressed.  PTAL, apologies this is 30 min late.

**Comment by derekwaynecarr** [comment]
I am also confused by Reviewable... do you know if we are going to turn it off by default now with github reviews being a thing?

**Comment by devin-donnelly** [comment]
I would prefer to use GitHub's PR Review system from here on out; it's cleaner than Reviewable.

**Comment by devin-donnelly** [comment]
Thanks for getting to this, @derekwaynecarr .

**Comment by timstclair** [comment]
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/admin/out-of-resource.md, line 223 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-223:-KRopeUdjoWbjruppUxI:b4wer22) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L223)):*
<details><summary><i>Previously, derekwaynecarr (Derek Carr) wrote…</i></summary>
> no, its supported for any resource under pressure, but it was new to 1.4 so got collapsed into this PR.  the heading depth confuses this, so i updated it, thanks.
</details>
I think the heading depth might still be wrong? Raise `Ranking pods for eviction` and the 2 subsections 1 level?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196)*

**Comment by vishh** [comment]
LGTM

**Comment by devin-donnelly** [comment]
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[docs/admin/out-of-resource.md, line 223 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196#-KREqC2yURKMP666d2Fa-r1-223:-KS2Y0XY0QJbcnURqfqI:b8p2ut5) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/7d44a8cdc10d641c83b3d27f3a29ed2a361468b6/docs/admin/out-of-resource.md#L223)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> I think the heading depth might still be wrong? Raise `Ranking pods for eviction` and the 2 subsections 1 level?
</details>
I think @timstclair is right--should be ### Ranking pods for eviction and #### With Imagefs and #### Without Imagefs .

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/1196)*

**Comment by devin-donnelly** [comment]
@derekwaynecarr , would you mind fixing this last comment in Reviewable about the header levels? Then I'll merge.

**Comment by derekwaynecarr** [comment]
@devin-donnelly -- restructured, headers should be good.

**Comment by devin-donnelly** [comment]
Looks good. Thanks! I'll merge it into 1.4.

**Pull request merged**